### ISSUE TITLE
Contructive critiques of the open source Letta project

### ISSUE THREAD

**Issue by distributev** [letta-ai/letta#2318]
If I am allowed, as a knowledgeable enough user who has actually tried to use Letta on 3 or 4 different occasions, these are the points where Letta can improve. This is something I am really looking to use, but every time, after insisting for 2 or 3 days to get it working, I end up giving up as it is "not yet ready".

1. Too many stack traces, many obscure ones, pop up. I am not sure if these are because of me, Letta, the LLM model used, the API provider (openrouter.ai), the environment (PowerShell console UTF8 configurations/support), etc. Last time, I consistently got two stack traces: one when Letta was trying to update its memory with some new information, which I believe was due to logging an "unsupported char" (misalignment between Letta's logging configuration and my PowerShell console related to encoding/UTF8 support); and the second stack trace came after an "empty" answer was returned by the LLM (perhaps because I stopped it as it took too long to respond). Once the empty message was added to the Letta database, all subsequent executions of Letta gave the same stack trace for all queries I tried. The solution was to manually remove the empty message from the sqlite.db messages table and remove the corresponding message ID from the agent's row in the database. 

2. Letta API changes from version to version. I got my create_agent.py script working fine with one Letta version, but when I tried the same script after 2 or 3 weeks, it failed due to API changes. I wasted another 1 or 2 days adapting the script to the new APIs, but I know this work is futile because in a month, I will need to do the same again.

3. Full ChatCompletions API support was advertised many times, but when I tried to use it and configure it, I found it undocumented. As I looked through the docs because I remembered it being advertised and mentioned previously, I saw that all mentions of it were removed from the docs. What is happening? Do they not want to support the ChatCompletions API anymore? Looking through the Letta code, I see some support for the ChatCompletions API through vLLM. But does this mean I need to start vLLM as well? Does vLLM work on a normal laptop as I have, or do I need expensive GPU support just to get a ChatCompletions API? There is also a docker-compose-vllm.yaml, which is probably needed to start vLLM on a GPU server.

4. ADE UI: This was working self-hosted and locally two months ago. Yesterday, I started the ADE "locally" again, and it forced me to redirect to the cloud version of the Letta app, where I was forced to create a cloud account with Letta. I then saw that even if I started the UI locally, it only proxied to use the cloud version. This goes against the open-source spirit that the initial MemGPT project and the later Letta project seemed to adhere to. 

5. No documentation exists about which LLMs work well with Letta and which do not. I wasted a lot of time trying all LLMs just to find out that only GPT-4o-mini was working fine (until recently), and yesterday, I tried the new Deepseek v3 and was pleasantly surprised that it worked too. Still, a "Letta LLM leaderboard" with all LLMs scored against Letta would be very valuable.

I hope you get this as positive criticisms, this is what I felt while spending a lot of time trying to get Letta working.

**Comment by cpacker** [comment]
Hey @distributev , thank you for the constructive feedback, really appreciate it (as well as all your other github issues / PRs you've raised in the past).

> Too many stack traces, many obscure ones, pop up. I am not sure if these are because of me, Letta, the LLM model used, the API provider (openrouter.ai), the environment (PowerShell console UTF8 configurations/support), etc

In this particular case - I would guess that OpenRouter may be a significant source of bugs. It's technically supported since you can use the OpenAI provider as a proxy but we don't test it.

Because Letta is built to be model agnostic and is pretty flexible with how you can connect it to LLM providers, it naturally has a pretty wide surface area as a piece of software. One source of friction that this has caused is that it makes it very hard for us as a small development team to make sure that all the "happy paths" for installation / usage are bug free.

For example - while you can technically use OpenRouter with Letta, there are a ton of models on OpenRouter that don't support function calling, and if they do, they often fail in weird ways that you wouldn't expect.

Another example of this friction is install paths. Yes, it's really nice to be able to install Letta with `pip` because it's really simple, but you can't install postgres with `pip`, which means that we have to maintain two different database backends, which lots of duplicate code (sqlite and postgres are not 1-1 swaps in their APIs).

In the past few months, we've started a new push to try and correct for some of these problems by narrowing our scope on what the "officially supported" ways to install and use Letta are. By narrowing our scope, we can free up dev time to focus on making sure the "official" ways to use Letta are bug free.

For example, we really recommend installing via docker now, since it's the simplest way to use postgres for most users. We also have modified the README to make clear that we are not supported database migrations when you use sqlite, because it's a ton of overhead to support alembic migration scripts for both postgres and sqlite separately - that dev time is better spent on adding new features and removing bugs on the primary install/usage paths.

Going back to the OpenRouter topic, I think it makes sense for us to also emphasize in the documentation that OpenRouter isn't really "officially" supported since many bugs may happen when you use it vs using model provider APIs directly.

> Letta API changes from version to version. I got my create_agent.py script working fine with one Letta version, but when I tried the same script after 2 or 3 weeks, it failed due to API changes. I wasted another 1 or 2 days adapting the script to the new APIs, but I know this work is futile because in a month, I will need to do the same again.

Totally understand the pain here. We're hoping the recent API changes are the last "big" changes since we did a big SDK migration - our Python and TS SDKs are now autogenerated off the REST API, which should also make the API much more stable. If you have an example file you wrote we're happy to help you translate it to the new version, just paste it here or on Discord.

> Full ChatCompletions API support was advertised many times, but when I tried to use it and configure it, I found it undocumented

A version of this is under development but it's not stable enough for us to document. Coming back to the top-level motivation of "focus on fewer usage paths and make the ones that are supported bug free", we only want to fully document that API when we feel like it's bug free and ready to go. There are a lot of important design decisions around how to implement a "proxy" API like that, so we don't want to release something then break the API a few weeks later.

> ADE UI: This was working self-hosted and locally two months ago. Yesterday, I started the ADE "locally" again, and it forced me to redirect to the cloud version of the Letta app

When we started designing the new ADE, we built it out to work with both self-deployed Letta servers and Letta Cloud. Right now while the ADE is in open beta, you use it with local Letta by pointing it at localhost but also any other address where you've self-deploying a server (your data is all stored on your local server, not the cloud).

Having multiple ways to use the ADE that are more natively integrated with the local server for local-only users is something we're actively looking into, but it's a significant dev time commitment so during the open beta period we're focusing on adding features to the ADE while it's accessible via the web app.

Re: the old ADE - because the API was changing significantly it wasn't really feasible for us to continue to spend dev cycles to keep patching the old version to keep up with the API changes. It's OSS software of course so you can always rollback your version to 0.5.0 to use it and it should work, but we just don't have the dev bandwidth to maintain two versions of the ADE separately.

> No documentation exists about which LLMs work well with Letta and which do not

This is something we are actively working on and hope to release very soon. In the meantime, at a high level I'd recommend gpt4o-mini from OpenAI, sonnet-3.5 from Anthropic, and llama 3.3 70b for open weights models, but our discord also has some model recommendations from users in the "models" section.


---

tl;dr We're trying to focus on fewer officially supported ways to install/use Letta, and make the officially support ways to use it as bug free and smooth as possible. Along these lines, my recommended install method is:

1. Use Docker to run the server (here I'm using `letta:nightly`):
```
docker run \
  -v ~/.letta/.persist/pgdata:/var/lib/postgresql/data \
  -p 8283:8283 \
  -e TOGETHER_API_KEY=... \
  -e ANTHROPIC_API_KEY=... \
  -e OPENAI_API_KEY=... \
  letta/letta:nightly
```

2. Don't use a proxy server (eg OpenRouter), use one of the officially supported/tested direct providers (Together, OpenAI, Anthropic, Ollama, etc)

Combining (1) and (2) will put you in our critical path for bug fixes (we are aggressively developing for this install/usage path, so if there are any bugs we'll catch them more quickly and be able to patch them faster).

**Comment by JCCPort** [comment]
@cpacker Would you be able to clarify the implication of the term "open beta period"? Is it planned that Letta ADE is to become a paid product?

**Comment by cpacker** [comment]
> Would you be able to clarify the implication of the term "open beta period"? Is it planned that Letta ADE is to become a paid product?

No sorry for the confusion - I meant "open beta" as in "early development (has bugs, user be warned!)", not in the sense of "free then paid". We have a managed service that runs agents in the cloud (in early access, you can request access via our website) - that is a paid service. But the ADE is still free - you can go to the same website as usual and just point the frontend to your own backend (local letta server). We also have a desktop variant (also quite early, less stable than the web version) you can use if you don't want to use a website.

**Comment by JCCPort** [comment]
@cpacker Thank you very much for the answer! I'm glad to hear that and look forward to getting started playing with Letta!
